### PR TITLE
support parenthesized expr in filters

### PR DESCRIPTION
**Pull request by xhh1989**
support parenthesized expr

### CODE REVIEW

File: go/vt/vtgate/planbuilder/expr.go [hunk 1]
```diff
@@ -35,6 +35,9 @@ func splitAndExpression(filters []sqlparser.Expr, node sqlparser.Expr) []sqlpars
 		filters = splitAndExpression(filters, node.Left)
 		return splitAndExpression(filters, node.Right)
 	}
+	if node, ok := node.(*sqlparser.ParenExpr); ok {
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/expr.go]: I don't think this will work for all use cases. For example, `a and (b or c)` will end up yielding `a and b or c` (because you're removing the parenthesis).

Comment (xhh1989) [reply go/vt/vtgate/planbuilder/expr.go]: "b or c"  is or expr and will not be splitted during the recursive

Comment (sougou) [reply go/vt/vtgate/planbuilder/expr.go]: The case I had in mind was this: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/planbuilder/select.go#L87.

In this place, I split the expressions into individual parts and just push them into the AST. At this point, the `b or c` would be unparenthesized. Fortunately, this eventually reaches this code: https://github.com/youtube/vitess/blob/master/go/vt/sqlparser/ast.go#L238, which always re-parenthesizes the RHS.

But we need to make sure that there are no other places where this gets blindly recombined in an AST without re-parenthesization.

In other words, the change is still risky. I'll need to look at all places where this is used and make sure nothing bad will happen.

Comment (sougou) [reply go/vt/vtgate/planbuilder/expr.go]: Correction: the `AddWhere` specifically re-parenthesizes `OR` expressions (not the RHS) as I mentioned before.

I must have done it to preserve precedence in case expressions were re-ordered. If so, it may be safe. But we still need proof that nothing else will break.

Comment (xhh1989) [reply go/vt/vtgate/planbuilder/expr.go]: Thanks for your patient reply, I encouter a problem,   for example I have a table messages with route key page,  sql like : delete from messages where page = 1 is ok, but delete from messages where (page = 1) is not supported (unsupported: multi-shard where clause in DML), I use third party client which generate sql like where (page=xxx) and (xxx) and ...., Is there any other methods to resolve this problem?

Comment (sougou) [reply go/vt/vtgate/planbuilder/expr.go]: Your change may just work. We just have to prove that it won't break anything. I'll need to look at all code paths and make sure nothing else will break. Should take me a couple of days.

Comment (xhh1989) [reply go/vt/vtgate/planbuilder/expr.go]: thanks very much, I will close the pull request, and open a issue in https://github.com/youtube/vitess/issues/3283
